src/data_types/tensor.hpp

```diff
@@ -0,0 +1,410 @@
+// SPDX-License-Identifier: MIT
+#pragma once
+#include <array>
+#include <cassert>
+
+#include <ddc/ddc.hpp>
+
+#include "ddc_aliases.hpp"
+#include "tensor_index_tools.hpp"
+#include "vector_index_tools.hpp"
+
+namespace detail {
+
+template <class ValidIndexSet>
+struct ToCoord;
+
+template <class... Dims>
+struct ToCoord<VectorIndexSet<Dims...>>
+{
+    using type = Coord<Dims...>;
+};
+
+template <class ValidIndexSet>
+using to_coord_t = typename ToCoord<ValidIndexSet>::type;
+
+} // namespace detail
+
+/**
+ * @brief A class representing a Tensor.
+ * @tparam ElementType The type of the elements of the tensor (usually double/complex).
+ * @tparam ValidIndexSet The indices that can be used along each dimension of the tensor.
+ */
+template <class ElementType, class... ValidIndexSet>
+class Tensor
+{
+    static_assert(std::conjunction_v<tensor_tools::is_vector_index_set<ValidIndexSet>...>);
+    static_assert(std::conjunction_v<std::disjunction<
+                          tensor_tools::is_covariant_vector_index_set<ValidIndexSet>,
+                          tensor_tools::is_contravariant_vector_index_set<ValidIndexSet>>...>);
```

Comment:
```suggestion
    static_assert((tensor_tools::is_vector_index_set<ValidIndexSet> && ...));
    static_assert(((tensor_tools::is_covariant_vector_index_set<ValidIndexSet> ||
                              tensor_tools::is_contravariant_vector_index_set<ValidIndexSet>) && ...));
```
Same as previous review, I would recommend classic C++ operators

Reply:
Fixed : 9396e804d6db52fae7a1514218fb1868a522ec3c